src/librustc_typeck/check/place_op.rs

```diff
@@ -246,15 +246,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
 
             match expr.kind {
                 hir::ExprKind::Index(ref base_expr, ref index_expr) => {
-                    // We need to get the final type in case dereferences were needed for the trait
-                    // to apply (#72002).
-                    let index_expr_ty = self.typeck_results.borrow().expr_ty_adjusted(index_expr);
```

Comment:
Ok, I narrowed down the issue to this line. This will cause the actual type (and lifetime) of index expression to be used, rather than a supertype of it.

Reply:
I am not sure if there is a way to retrieve the input_ty in try_index_step. If we could we could just use it as index_expr_ty.